openmsupply/application-manager-server
# #658 Snapshot migration

Implements #658, using Solution 1 outlined in that issue. It's not a huge PR so can change if there's a problem with this approach.

Seems to be working okay though.

- exported .sql script is just one file (concatenation of all script files). Found a couple of missing ";" in existing scripts as a result
- manually calls the "migrate" script at the end (which deals with [this comment](https://github.com/openmsupply/application-manager-server/issues/658#issuecomment-1042277402))
- updated the "database_init" command so it loads the snapshot via the "useSnapshot.ts" script (which is should have been doing anyway, since it has a few extra little things)
- modified "initialise_database.sh" so take an optional parameter -- the name of the ".sql" script file to initialise with. If the snapshot being loaded has a script file, the filepath is passed to this script by "useSnapshot", if not it just runs the default .sql scripts
- updated some of the "execSync" calls to add " { stdio: 'inherit' } " option -- this prints the subshell's console output to main console (was previously not shown)
- tweak "migrateData":
  - add "CREATE TABLE system_info"
  - a generic "changeSchema" database method for schema changing queries (no variables), with appropriate error handling (doesn't throw, just console.log)

### Testing

- Try testing both using the "database_init" command line as well as via the "Snapshots" ui in browser
- Try loading old snapshots and note that the schema is migrated after loading (it won't work with snapshots that have incompatible schemas though, i.e. trying to insert to a field/table that no longer exists -- we can't do anything about these, will need to be manually updated if there are any and we still need to keep them)
- Export a snapshot. Change something in the schema (e.g. rename a field, delete a table). Reload snapshot, and note the schema will be put back to how it was when the snapshot was saved
- Change something in the main database .sql scripts. Load a snapshot that has a build script saved with it and note that the snapshot's schema is loaded, not the modified one (obviously we would have migration code to bring the snapshot schema up-to-date with the main schema script)
- Note that after any snapshot load, the "system_info" table should a) exist; and b) have a newly-created record with the current version in it
- Anything else? Find any thing that breaks or doesn't behave as expected?

**Edit**: TO-DO:

- ~~as per @andreievg 's comment -- I'm going to add a safeguard to forbid loading a snapshot with a higher version number than the current installation.~~ Done 35112aa1
- ~~add migration code for trigger/function modified [here](https://github.com/openmsupply/application-manager-server/pull/660)~~ Done 949bc078
